### DISCUSSION TITLE
Wemos D1 mini V4.0.0 supported?

### DISCUSSION THREAD

**Discussion by GroteGehaktBal** [Kaldek/rat-ratgdo#111] [Q&A]
I have a [wemos D1 mini V4.0.0](https://www.wemos.cc/en/latest/d1/d1_mini.html) (The one with USB C)

I was wondering if this board is supported as I don't see the D pins on it but instead it has pins like 5, 4, 0, 2.
Are those pins just renamed and will it work or do I need an older version of the D1 mini?

**Comment by foreverimagining** [comment]
So they have enough GPIOs available since you only need 3 for a basic Security+1.0/2.0 setup or 8 if you wanted to have the dry contact side of the board as well. Those pins are available GPIOs like the ones usually labeled with `D`. Looking at the schematic for those boards, you would probably use `GPIO#` to assign the pins to the various functions when configuring your ESPHome firmware as the GPIO numbers match with the printed numbers. Not all boards actually allow for assigning pins with `D#`, and when they do, the `D` numbers don't usually match the GPIO numbers that they stand for.

Where the problem might come in is whether that particular board is currently supported by ESPHome, or if there is a near enough equivalent that you could get by. 

You see, when you configure the firmware, you have to assign a board so that it will know a little something about the hardware that you are using, including the particular microcontroller and how much flash is available. This actually led to an issue with RATGDOs where the firmware was perceived to be too big for the available flash, which was considered less than reality because of the board that was chosen to be assigned. The problem was noticed when it was preventing people from updating, and the ultimate solution was a slight tweak to the board selection.

These boards are running an ESP-8266EX microcontroller, which I assume would fall under the [ESP8266 platform](https://esphome.io/components/esp8266) seeing as they do use the datasheet as a source near the bottom. However, a cursory search through the [board list](https://registry.platformio.org/platforms/platformio/espressif8266/boards) didn't seem to turn up these specific boards, which may mean that you will have to hunt through for one with the same general stats if you want to try it. 

You could look around to see if anyone has used these boards in general with ESPHome and what their configurations look like to see what board they assign. If they do work, then it should be possible to reconfigure the RATGDO firmware a little to use with these boards.

**Reply by GroteGehaktBal** [reply]
Thank you for your reply, I am currently trying to make this work and I connected it like the schematic and installed 2.5x firmware. So I connected it using GPIO13(MOSI), SCL and SDA on the wemos d1 mini V4 and my garage door opener has these ports like the one on this picture (not my picture)![image](https://github.com/user-attachments/assets/d1c8feb8-d732-416f-b17a-4d95518ef2f6)

I have connected the wires like in that picture, so red to 0, white to 2 and black to 3.
This is not yet working because I get this:
```Not connected to GDO, ignoring command```

What can I try to do now? Or should I just get a different esp?

**Comment by GroteGehaktBal** [comment]
I think I would only have to change the code so that it uses gpio 5 for SCL and gpio 4 for SDA. That would probably fix it but I have no idea on how to do that or if that is even possible.

**Reply by foreverimagining** [reply]
It's actually surprisingly easy to modify ESPHome configs, but I'll admit that it's a little hard to work out without reading basically all of their documentation.

In order to reassign the pins, you will need to take a look at the basic configuration files in [this folder](https://github.com/ratgdo/esphome-ratgdo/tree/main/static) such as [this one](https://github.com/ratgdo/esphome-ratgdo/blob/main/static/v25board_esp8266_d1_mini.yaml) for the v2.5 with an ESP8266 D1 Mini. Which one you use depends as much on the protocol you need (Security+1.0 vs. 2.0 vs. dry contact) as the microcontroller that you are using. 

At the top of the files, you have the `substitutions` section, which is a feature designed to make it easier to modify commonly changed aspects of the overall code.
```
substitutions:
  id_prefix: ratgdov25
  friendly_name: "ratgdov2.5"
  uart_tx_pin: D1
  uart_rx_pin: D2
  input_obst_pin: D7
  status_door_pin: D0
  status_obstruction_pin: D8
  dry_contact_open_pin: D5
  dry_contact_close_pin: D6
  dry_contact_light_pin: D3
```
In here, you can reassign the pins being used for the TX, RX, obstruction sensors, etc. by simply changing the pin like so:
```
uart_tx_pin: GPIO1
```
or even replace the pins with nothing by assigning `none` in case you wanted to use them for something else (like in my case, an ultrasonic sensor) because each pin can only be assigned once or the firmware won't compile:
```
dry_contact_open_pin: none
```
Since it looked like you can't use the `D#` nomenclature with the Lolin D1 mini, the microcontroller will probably not have any of the default pins actually assigned to any of the GPIOs, which might be why it's saying that you aren't connected to the GDO.

### Bonus
You can also remove sensors that you aren't going to use to have fewer entities created in HA or de-clutter the web server. You just have to go through the [base.yaml](https://github.com/ratgdo/esphome-ratgdo/blob/main/base.yaml) to find the particular sensor you don't want, then use something like the following to remove them:
```
binary_sensor:
  - id: !remove ${id_prefix}_dry_contact_open
```
Just remember, you have to use the component's section header, then `  -id: !remove`, then the name. You may recognize the `${id_prefix}` variable from the `substitutions` section.

## And in case you didn't already know:
This process is based on compiling a custom firmware by creating a `New Device` in ESPHome in Home Assistant rather than using the RATGDO web installer. You would just copy the contents of a basic config file from the `static` folder I linked to above, make your changes, then compile the firmware, and flash it. You can also adopt your already-flashed device into ESPHome to make your edits. In which case, I want to say you don't have to copy the whole config file, just add the necessary substitutions and any other changes, but it's been a while so my memory is a little foggy.

**Reply by Guy559** [reply]
@GroteGehaktBal , did you get this working?

I have my original RATGDOs working with ESP8266. Looking to move to ESP32... trying to determine which hardware (ESP32s) work with native software...

**Reply by GroteGehaktBal** [reply]
@Guy559 Yes I did get it working, the only problem I have is that it doesn't always show the state correctly (if it is open or closed) and then the close and open buttons don't always work properly. So I just use the toggle door button which does always work.
